File: algorithms/hk-re-vv/cpp/hk_re_vv.cpp
/*
 * Hopcroft-Karp Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * VV (vector-of-vectors) adjacency.
 *
 * Refactored to separate input (BipartiteGraph), output (BipartiteMatching),
 * and algorithm state (HKState).
 */

#include <cstdio>
#include <cstdlib>
#include <cstdint>
#include <vector>
#include <algorithm>
#include <string>
#include <chrono>
#include <climits>

constexpr int32_t NIL = -1;

/* ---------- Input: BipartiteGraph ---------- */

struct BipartiteGraph {
    size_t sNumVtxs;
    size_t tNumVtxs;
    size_t numEdgs;
    std::vector<std::vector<int32_t>> sAdj;  // sAdj[u] = t-neighbors of s-vertex u
    std::vector<std::vector<int32_t>> tAdj;  // tAdj[v] = s-neighbors of t-vertex v
};

/* ---------- Output: BipartiteMatching ---------- */

struct BipartiteMatching {
    size_t sNumVtxs;
    size_t tNumVtxs;
    size_t numEdgs;
    std::vector<int32_t> sMate;   // sMate[u] = t-partner of s-vertex u, or NIL
    std::vector<int32_t> tMate;   // tMate[v] = s-partner of t-vertex v, or NIL
};

/* ---------- State: HKState ---------- */

struct HKState {
    std::vector<int32_t> dist;    // length sNumVtxs+1; dist[sNumVtxs] is the NIL sentinel
};

/* ---------- BipartiteGraph construction ---------- */

BipartiteGraph buildBipartiteGraph(size_t sNumVtxs, size_t tNumVtxs,
                                   const std::vector<std::pair<int32_t,int32_t>>& edges) {
    BipartiteGraph g;
    g.sNumVtxs = sNumVtxs;
    g.tNumVtxs = tNumVtxs;
    g.sAdj.resize(sNumVtxs);
    g.tAdj.resize(tNumVtxs);
    for (auto& e : edges) {
        int32_t u = e.first, v = e.second;
        if (u >= 0 && (size_t)u < sNumVtxs && v >= 0 && (size_t)v < tNumVtxs) {
            g.sAdj[u].push_back(v);
            g.tAdj[v].push_back(u);
        }
    }
    for (size_t u = 0; u < sNumVtxs; u++) {
        std::sort(g.sAdj[u].begin(), g.sAdj[u].end());
        g.sAdj[u].erase(std::unique(g.sAdj[u].begin(), g.sAdj[u].end()), g.sAdj[u].end());
    }
    for (size_t v = 0; v < tNumVtxs; v++) {
        std::sort(g.tAdj[v].begin(), g.tAdj[v].end());
        g.tAdj[v].erase(std::unique(g.tAdj[v].begin(), g.tAdj[v].end()), g.tAdj[v].end());
    }
    size_t total = 0;
    for (size_t u = 0; u < sNumVtxs; u++) total += g.sAdj[u].size();
    g.numEdgs = total;
    return g;
}

/* ---------- BipartiteMatching construction ---------- */

BipartiteMatching emptyBipartiteMatching(const BipartiteGraph& g) {
    BipartiteMatching m;
    m.sNumVtxs = g.sNumVtxs;
    m.tNumVtxs = g.tNumVtxs;
    m.numEdgs = 0;
    m.sMate.assign(g.sNumVtxs, NIL);
    m.tMate.assign(g.tNumVtxs, NIL);
    return m;
}

/* ---------- Greedy initial matching: simple ---------- */

int32_t greedyInit(const BipartiteGraph& g, BipartiteMatching& m) {
    int32_t cnt = 0;
    for (size_t u = 0; u < g.sNumVtxs; u++) {
        if (m.sMate[u] != NIL) continue;
        for (int32_t v : g.sAdj[u]) {
            if (m.tMate[v] == NIL) {
                m.sMate[u] = v;
                m.tMate[v] = (int32_t)u;
                cnt++;
                break;
            }
        }
    }
    m.numEdgs += cnt;
    return cnt;
}

/* ---------- Greedy initial matching: min-degree ---------- */

int32_t greedyInitMd(const BipartiteGraph& g, BipartiteMatching& m) {
    int32_t cnt = 0;
    std::vector<int32_t> deg(g.tNumVtxs, 0);
    for (size_t u = 0; u < g.sNumVtxs; u++)
        for (int32_t v : g.sAdj[u])
            deg[v]++;

    std::vector<int32_t> order(g.sNumVtxs);
    for (size_t i = 0; i < g.sNumVtxs; i++) order[i] = (int32_t)i;
    std::sort(order.begin(), order.end(), [&](int32_t a, int32_t b){
        return g.sAdj[a].size() < g.sAdj[b].size() ||
               (g.sAdj[a].size() == g.sAdj[b].size() && a < b);
    });

    for (int32_t u : order) {
        if (m.sMate[u] != NIL) continue;
        int32_t best = NIL, bestDeg = INT_MAX;
        for (int32_t v : g.sAdj[u]) {
            if (m.tMate[v] == NIL && deg[v] < bestDeg) {
                best = v;
                bestDeg = deg[v];
            }
        }
        if (best != NIL) {
            m.sMate[u] = best;
            m.tMate[best] = u;
            cnt++;
        }
    }
    m.numEdgs += cnt;
    return cnt;
}

/* ---------- HK BFS ---------- */

static bool bfs(const BipartiteGraph& g, const BipartiteMatching& m, HKState& s) {
    std::vector<int32_t> queue(g.sNumVtxs);
    int32_t qh = 0, qt = 0;

    for (size_t u = 0; u < g.sNumVtxs; u++) {
        if (m.sMate[u] == NIL) { s.dist[u] = 0; queue[qt++] = (int32_t)u; }
        else s.dist[u] = INT_MAX;
    }
    s.dist[g.sNumVtxs] = INT_MAX;  /* NIL sentinel */

    while (qh < qt) {
        int32_t u = queue[qh++];
        if (s.dist[u] < s.dist[g.sNumVtxs]) {
            for (int32_t v : g.sAdj[u]) {
                int32_t pn = (m.tMate[v] == NIL) ? (int32_t)g.sNumVtxs : m.tMate[v];
                if (s.dist[pn] == INT_MAX) {
                    s.dist[pn] = s.dist[u] + 1;
                    if (m.tMate[v] != NIL) queue[qt++] = m.tMate[v];
                }
            }
        }
    }
    return s.dist[g.sNumVtxs] != INT_MAX;
}

/* ---------- HK DFS ---------- */

static bool dfs(int32_t u, const BipartiteGraph& g, BipartiteMatching& m, HKState& s) {
    if (u == NIL) return true;
    for (int32_t v : g.sAdj[u]) {
        int32_t pn = (m.tMate[v] == NIL) ? (int32_t)g.sNumVtxs : m.tMate[v];
        if (s.dist[pn] == s.dist[u] + 1) {
            if (dfs(m.tMate[v], g, m, s)) {
                m.tMate[v] = u;
                m.sMate[u] = v;
                return true;
            }
        }
    }
    s.dist[u] = INT_MAX;
    return false;
}

/* ---------- Top-level Hopcroft-Karp ---------- */

void hopcroftKarp(const BipartiteGraph& g, BipartiteMatching& m) {
    HKState s;
    s.dist.assign(g.sNumVtxs + 1, 0);

    int32_t phases = 0;
    int32_t newEdgs = 0;
    while (bfs(g, m, s)) {
        phases++;
        for (size_t u = 0; u < g.sNumVtxs; u++) {
            if (m.sMate[u] == NIL && dfs((int32_t)u, g, m, s)) {
                newEdgs++;
            }
        }
    }
    m.numEdgs += newEdgs;
    printf("Phases: %d\n", phases);
}

/* ---------- Validation ---------- */

void validateBipartiteMatching(const BipartiteGraph& g, const BipartiteMatching& m) {
    int32_t errors = 0;
    int32_t matchedS = 0, matchedT = 0;

    for (size_t u = 0; u < g.sNumVtxs; u++) {
        if (m.sMate[u] != NIL) {
            matchedS++;
            int32_t v = m.sMate[u];
            if (v < 0 || (size_t)v >= g.tNumVtxs) {
                fprintf(stderr, "ERROR: sMate[%zu] = %d out of range\n", u, v);
                errors++;
            } else if (m.tMate[v] != (int32_t)u) {
                fprintf(stderr, "ERROR: sMate[%zu]=%d but tMate[%d]=%d\n", u, v, v, m.tMate[v]);
                errors++;
            } else if (!std::binary_search(g.sAdj[u].begin(), g.sAdj[u].end(), v)) {
                fprintf(stderr, "ERROR: edge (%zu,%d) not in graph\n", u, v);
                errors++;
            }
        }
    }
    for (size_t v = 0; v < g.tNumVtxs; v++) {
        if (m.tMate[v] != NIL) matchedT++;
    }

    printf("\n=== Validation Report ===\n");
    printf("Matching size: %zu\n", m.numEdgs);
    printf("S matched: %d, T matched: %d\n", matchedS, matchedT);
    printf("%s\n", errors > 0 ? "VALIDATION FAILED" : "VALIDATION PASSED");
    printf("=========================\n\n");
}

/* ---------- Main ---------- */

int main(int argc, char* argv[]) {
    printf("Hopcroft-Karp Algorithm - C++ Implementation (VV)\n");
    printf("===================================================\n\n");

    if (argc < 2) { printf("Usage: %s <filename> [--greedy|--greedy-md]\n", argv[0]); return 1; }
    int greedyMode = 0;
    for (int i = 2; i < argc; i++) {
        if (std::string(argv[i]) == "--greedy") greedyMode = 1;
        else if (std::string(argv[i]) == "--greedy-md") greedyMode = 2;
    }

    FILE* f = fopen(argv[1], "r");
    if (!f) { fprintf(stderr, "Cannot open file: %s\n", argv[1]); return 1; }

    int sNumVtxs, tNumVtxs, numEdgsIn;
    if (fscanf(f, "%d %d %d", &sNumVtxs, &tNumVtxs, &numEdgsIn) != 3) {
        fprintf(stderr, "Bad header\n"); fclose(f); return 1;
    }

    std::vector<std::pair<int32_t,int32_t>> edges;
    edges.reserve(numEdgsIn);
    for (int i = 0; i < numEdgsIn; i++) {
        int u, v;
        if (fscanf(f, "%d %d", &u, &v) != 2) break;
        edges.push_back({(int32_t)u, (int32_t)v});
    }
    fclose(f);

    printf("Graph: %d s-vertices, %d t-vertices, %zu edges\n",
           sNumVtxs, tNumVtxs, edges.size());

    BipartiteGraph bipartiteGraph = buildBipartiteGraph(sNumVtxs, tNumVtxs, edges);
    BipartiteMatching bipartiteMatching = emptyBipartiteMatching(bipartiteGraph);

    auto t0 = std::chrono::high_resolution_clock::now();

    int32_t greedySize = 0;
    if (greedyMode == 1) greedySize = greedyInit(bipartiteGraph, bipartiteMatching);
    else if (greedyMode == 2) greedySize = greedyInitMd(bipartiteGraph, bipartiteMatching);

    hopcroftKarp(bipartiteGraph, bipartiteMatching);

    auto t1 = std::chrono::high_resolution_clock::now();

    validateBipartiteMatching(bipartiteGraph, bipartiteMatching);

    printf("Matching size: %zu\n", bipartiteMatching.numEdgs);
    if (greedyMode > 0) {
        printf("Greedy init size: %d\n", greedySize);
        if (bipartiteMatching.numEdgs > 0)
            printf("Greedy/Final: %.2f%%\n", 100.0 * greedySize / bipartiteMatching.numEdgs);
        else
            printf("Greedy/Final: NA\n");
    }
    printf("Time: %ld ms\n",
           (long)std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count());
    return 0;
}

File: algorithms/hk-re-vv/rs/hk_re_vv.rs
/*
 * Hopcroft-Karp Algorithm - O(E√V) Maximum Bipartite Matching
 *
 * VV (vector-of-vectors) adjacency.
 *
 * Refactored to separate input (BipartiteGraph), output (BipartiteMatching),
 * and algorithm state (HKState). The decoupling lets us drop the .clone()
 * calls on the inner adjacency that the original (single-struct) Rust port
 * needed to satisfy the borrow checker.
 */

use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;

const NIL: i32 = -1;

/* ---------- Input: BipartiteGraph ---------- */

#[allow(dead_code)]
struct BipartiteGraph {
    s_num_vtxs: usize,
    t_num_vtxs: usize,
    num_edgs: usize,
    s_adj: Vec<Vec<i32>>,    // s_adj[u] = t-neighbors of s-vertex u
    t_adj: Vec<Vec<i32>>,    // t_adj[v] = s-neighbors of t-vertex v
}

/* ---------- Output: BipartiteMatching ---------- */

#[allow(dead_code)]
struct BipartiteMatching {
    s_num_vtxs: usize,
    t_num_vtxs: usize,
    num_edgs: usize,
    s_mate: Vec<i32>,        // s_mate[u] = t-partner of s-vertex u, or NIL
    t_mate: Vec<i32>,        // t_mate[v] = s-partner of t-vertex v, or NIL
}

/* ---------- State: HKState ---------- */

struct HKState {
    dist: Vec<i32>,          // length s_num_vtxs+1; dist[s_num_vtxs] is the NIL sentinel
}

/* ---------- BipartiteGraph construction ---------- */

fn build_bipartite_graph(s_num_vtxs: usize, t_num_vtxs: usize,
                         edges: &[(usize, usize)]) -> BipartiteGraph {
    let mut s_adj: Vec<Vec<i32>> = vec![Vec::new(); s_num_vtxs];
    let mut t_adj: Vec<Vec<i32>> = vec![Vec::new(); t_num_vtxs];
    for &(u, v) in edges {
        if u < s_num_vtxs && v < t_num_vtxs {
            s_adj[u].push(v as i32);
            t_adj[v].push(u as i32);
        }
    }
    for adj in &mut s_adj { adj.sort_unstable(); adj.dedup(); }
    for adj in &mut t_adj { adj.sort_unstable(); adj.dedup(); }

    let num_edgs: usize = s_adj.iter().map(|a| a.len()).sum();

    BipartiteGraph {
        s_num_vtxs,
        t_num_vtxs,
        num_edgs,
        s_adj,
        t_adj,
    }
}

/* ---------- BipartiteMatching construction ---------- */

fn empty_bipartite_matching(g: &BipartiteGraph) -> BipartiteMatching {
    BipartiteMatching {
        s_num_vtxs: g.s_num_vtxs,
        t_num_vtxs: g.t_num_vtxs,
        num_edgs: 0,
        s_mate: vec![NIL; g.s_num_vtxs],
        t_mate: vec![NIL; g.t_num_vtxs],
    }
}

/* ---------- Greedy initial matching: simple ---------- */

fn greedy_init(g: &BipartiteGraph, m: &mut BipartiteMatching) -> usize {
    let mut cnt: usize = 0;
    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] != NIL { continue; }
        for &v in &g.s_adj[u] {
            if m.t_mate[v as usize] == NIL {
                m.s_mate[u] = v;
                m.t_mate[v as usize] = u as i32;
                cnt += 1;
                break;
            }
        }
    }
    m.num_edgs += cnt;
    cnt
}

/* ---------- Greedy initial matching: min-degree ---------- */

fn greedy_init_md(g: &BipartiteGraph, m: &mut BipartiteMatching) -> usize {
    let mut cnt: usize = 0;
    let mut deg = vec![0usize; g.t_num_vtxs];
    for u in 0..g.s_num_vtxs {
        for &v in &g.s_adj[u] {
            deg[v as usize] += 1;
        }
    }

    let mut order: Vec<usize> = (0..g.s_num_vtxs).collect();
    order.sort_unstable_by(|&a, &b| {
        g.s_adj[a].len().cmp(&g.s_adj[b].len()).then(a.cmp(&b))
    });

    for u in order {
        if m.s_mate[u] != NIL { continue; }
        let mut best: i32 = NIL;
        let mut best_deg = usize::MAX;
        for &v in &g.s_adj[u] {
            if m.t_mate[v as usize] == NIL && deg[v as usize] < best_deg {
                best = v;
                best_deg = deg[v as usize];
            }
        }
        if best != NIL {
            m.s_mate[u] = best;
            m.t_mate[best as usize] = u as i32;
            cnt += 1;
        }
    }
    m.num_edgs += cnt;
    cnt
}

/* ---------- HK BFS ---------- */

fn bfs(g: &BipartiteGraph, m: &BipartiteMatching, s: &mut HKState) -> bool {
    let mut queue: Vec<usize> = Vec::with_capacity(g.s_num_vtxs);

    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] == NIL {
            s.dist[u] = 0;
            queue.push(u);
        } else {
            s.dist[u] = i32::MAX;
        }
    }
    s.dist[g.s_num_vtxs] = i32::MAX;

    let mut qi = 0;
    while qi < queue.len() {
        let u = queue[qi];
        qi += 1;
        if s.dist[u] < s.dist[g.s_num_vtxs] {
            for &v in &g.s_adj[u] {
                let paired = if m.t_mate[v as usize] == NIL {
                    g.s_num_vtxs
                } else {
                    m.t_mate[v as usize] as usize
                };
                if s.dist[paired] == i32::MAX {
                    s.dist[paired] = s.dist[u] + 1;
                    if m.t_mate[v as usize] != NIL {
                        queue.push(m.t_mate[v as usize] as usize);
                    }
                }
            }
        }
    }
    s.dist[g.s_num_vtxs] != i32::MAX
}

/* ---------- HK DFS ---------- */

fn dfs(u_opt: i32, g: &BipartiteGraph, m: &mut BipartiteMatching, s: &mut HKState) -> bool {
    if u_opt == NIL { return true; }
    let u = u_opt as usize;

    /* No clone: g.s_adj is borrowed via &g, m.t_mate/s_mate via &mut m -- disjoint borrows. */
    for i in 0..g.s_adj[u].len() {
        let v = g.s_adj[u][i];
        let paired = if m.t_mate[v as usize] == NIL {
            g.s_num_vtxs as i32
        } else {
            m.t_mate[v as usize]
        };
        if s.dist[paired as usize] == s.dist[u] + 1 {
            if dfs(m.t_mate[v as usize], g, m, s) {
                m.t_mate[v as usize] = u as i32;
                m.s_mate[u] = v;
                return true;
            }
        }
    }
    s.dist[u] = i32::MAX;
    false
}

/* ---------- Top-level Hopcroft-Karp ---------- */

fn hopcroft_karp(g: &BipartiteGraph, m: &mut BipartiteMatching) {
    let mut s = HKState {
        dist: vec![0i32; g.s_num_vtxs + 1],
    };

    let mut phases: i32 = 0;
    let mut new_edgs: usize = 0;
    while bfs(g, m, &mut s) {
        phases += 1;
        for u in 0..g.s_num_vtxs {
            if m.s_mate[u] == NIL && dfs(u as i32, g, m, &mut s) {
                new_edgs += 1;
            }
        }
    }
    m.num_edgs += new_edgs;
    println!("Phases: {}", phases);
}

/* ---------- Validation ---------- */

fn validate_bipartite_matching(g: &BipartiteGraph, m: &BipartiteMatching) {
    let mut errors = 0;
    let mut matched_s = 0;
    let mut matched_t = 0;

    for u in 0..g.s_num_vtxs {
        if m.s_mate[u] != NIL {
            matched_s += 1;
            let v = m.s_mate[u];
            if v < 0 || (v as usize) >= g.t_num_vtxs {
                eprintln!("ERROR: s_mate[{}] = {} out of range", u, v);
                errors += 1;
            } else if m.t_mate[v as usize] != u as i32 {
                eprintln!("ERROR: s_mate[{}]={} but t_mate[{}]={}", u, v, v, m.t_mate[v as usize]);
                errors += 1;
            } else if !g.s_adj[u].binary_search(&v).is_ok() {
                eprintln!("ERROR: edge ({},{}) not in graph", u, v);
                errors += 1;
            }
        }
    }
    for v in 0..g.t_num_vtxs {
        if m.t_mate[v] != NIL { matched_t += 1; }
    }

    println!("\n=== Validation Report ===");
    println!("Matching size: {}", m.num_edgs);
    println!("S matched: {}, T matched: {}", matched_s, matched_t);
    println!("{}", if errors > 0 { "VALIDATION FAILED" } else { "VALIDATION PASSED" });
    println!("=========================\n");
}

/* ---------- Graph loader ---------- */

fn load_graph(filename: &str) -> Result<(usize, usize, Vec<(usize, usize)>), Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    let mut lines = reader.lines();

    let first = lines.next().ok_or("Empty file")??;
    let parts: Vec<&str> = first.split_whitespace().collect();
    if parts.len() != 3 {
        return Err("First line must have 3 numbers".into());
    }
    let s_num_vtxs: usize = parts[0].parse()?;
    let t_num_vtxs: usize = parts[1].parse()?;
    let m: usize = parts[2].parse()?;

    let mut edges = Vec::with_capacity(m);
    for line in lines {
        let line = line?;
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 {
            let u: usize = parts[0].parse()?;
            let v: usize = parts[1].parse()?;
            edges.push((u, v));
        }
    }
    Ok((s_num_vtxs, t_num_vtxs, edges))
}

/* ---------- Main ---------- */

fn main() {
    println!("Hopcroft-Karp Algorithm - Rust Implementation (VV)");
    println!("====================================================\n");

    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <filename> [--greedy|--greedy-md]", args[0]);
        std::process::exit(1);
    }
    let greedy_mode: i32 = if args.iter().any(|a| a == "--greedy-md") { 2 }
                          else if args.iter().any(|a| a == "--greedy") { 1 }
                          else { 0 };

    match load_graph(&args[1]) {
        Ok((s_num_vtxs, t_num_vtxs, edges)) => {
            println!("Graph: {} s-vertices, {} t-vertices, {} edges",
                     s_num_vtxs, t_num_vtxs, edges.len());

            let bipartite_graph = build_bipartite_graph(s_num_vtxs, t_num_vtxs, &edges);
            let mut bipartite_matching = empty_bipartite_matching(&bipartite_graph);

            let start = Instant::now();

            let greedy_size: usize = match greedy_mode {
                1 => greedy_init(&bipartite_graph, &mut bipartite_matching),
                2 => greedy_init_md(&bipartite_graph, &mut bipartite_matching),
                _ => 0,
            };

            hopcroft_karp(&bipartite_graph, &mut bipartite_matching);

            let duration = start.elapsed();

            validate_bipartite_matching(&bipartite_graph, &bipartite_matching);

            println!("Matching size: {}", bipartite_matching.num_edgs);
            if greedy_mode > 0 {
                let fs = bipartite_matching.num_edgs;
                println!("Greedy init size: {}", greedy_size);
                if fs > 0 {
                    println!("Greedy/Final: {:.2}%", 100.0 * greedy_size as f64 / fs as f64);
                } else {
                    println!("Greedy/Final: NA");
                }
            }
            println!("Time: {} ms", duration.as_millis());
        }
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    }
}
